networkit/cpp/independentset/Luby.cpp

```diff
@@ -41,30 +47,30 @@ std::vector<bool> Luby::run(const Graph &G) {
 
     while (!empty()) {
         // choose set S - weighted choice of active nodes with probability $1 / 2 \omega(v)$
-        std::vector<bool> S(G.numberOfNodes(), false);
+        std::vector<std::atomic<bool>> S(G.numberOfNodes());
+
         G.parallelForNodes([&](node u) {
             if (V[u]) {
                 if (Aux::Random::probability() < nodeProbability(u)) {
                     S[u] = true; // add node to S
                 }
             }
         });
+
         // remove non-independent nodes from S to get S'
-        G.parallelForEdges([&](node u, node v) {
-            if (u != v) {           // exclude self-loops
-                if (S[u] && S[v]) { // u and v are not independent (note: S is subset of V')
-                    // remove node with smaller degree
-                    edgeweight wu = weightedDegree(u);
-                    edgeweight wv = weightedDegree(v);
-                    if (wu > wv) {
-                        S[v] = false;
-                    } else if (wv > wu) {
-                        S[u] = false;
-                    } else {          // tie
-                        S[v] = false; // arbitrary decision
+        G.parallelForNodes([&](node u) {
```

Comment:
While the code works on CI, it is not guaranteed to work due to race conditions. The atomicity of bool and `std::atomic` helps when doing atomic operations like read/write. So it might happen, that after checking for `(S[u] && S[v])`, another thread is already setting `S[v]` or `S[u]` to `false` - leading to wrong results.

I think there is no way around using a critical section:

```
        std::vector<bool> S(G.numberOfNodes());
        ...
        // remove non-independent nodes from S to get S'
        G.parallelForNodes([&](node u) {
            if (S[u]) {
                G.forNeighborsOf(u, [&](node v) {
                    if (u < v) { // exclude self-loops and enforce strict order
#pragma omp critical
                        {
                            // remove node with smaller degree from S
                            if (S[u] && S[v]) {
                                if (weightedDegree(u) > weightedDegree(v)) {
                                    S[v] = false;
                                } else {
                                    S[u] = false;
                                }
                            }
                        }
                    }
                });
            }
        });
```

While the code is slower, it is not worsened too much (benchmarked with 8 cores):

```
# Before (w. race conditions): 
❯ python benchmark_luby.py
Time:  0.018196725845336915 +/- 0.001036126555402532

# After (w. critical section):
❯ python benchmark_luby.py
Time:  0.05442938804626465 +/- 0.013922199304217321